libkbfs/block_journal.go

```diff
@@ -372,141 +219,20 @@ func (j *blockJournal) end() (journalOrdinal, error) {
-	}
-
-	if j.refs[id] == nil {
-		j.refs[id] = make(blockRefMap)
-	}
-
-	return j.refs[id].put(refEntry.context, refEntry.status, ordinal)
-}
-
-func (j *blockJournal) exists(id BlockID) error {
-	_, err := os.Stat(j.blockDataPath(id))
-	return err
-}
-
-func (j *blockJournal) getDataSize(id BlockID) (int64, error) {
-	fi, err := os.Stat(j.blockDataPath(id))
-	if err != nil {
-		return 0, err
-	}
-	return fi.Size(), nil
-}
-
-func (j *blockJournal) getData(id BlockID) (
-	[]byte, kbfscrypto.BlockCryptKeyServerHalf, error) {
-	data, err := ioutil.ReadFile(j.blockDataPath(id))
-	if os.IsNotExist(err) {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{},
-			blockNonExistentError{id}
-	} else if err != nil {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
-	}
-
-	keyServerHalfPath := j.keyServerHalfPath(id)
-	buf, err := ioutil.ReadFile(keyServerHalfPath)
-	if os.IsNotExist(err) {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{},
-			blockNonExistentError{id}
-	} else if err != nil {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
-	}
-
-	// Check integrity.
-
-	dataID, err := j.crypto.MakePermanentBlockID(data)
-	if err != nil {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
-	}
-
-	if id != dataID {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, fmt.Errorf(
-			"Block ID mismatch: expected %s, got %s", id, dataID)
-	}
-
-	var serverHalf kbfscrypto.BlockCryptKeyServerHalf
-	err = serverHalf.UnmarshalBinary(buf)
-	if err != nil {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
-	}
-
-	return data, serverHalf, nil
+func (j *blockJournal) hasData(id BlockID) error {
+	return j.s.hasData(id)
 }
 
 // All functions below are public functions.
 
-func (j *blockJournal) hasRef(id BlockID) bool {
-	return j.refs[id] != nil
-}
-
-func (j *blockJournal) hasNonArchivedRef(id BlockID) bool {
-	refs := j.refs[id]
-	return (refs != nil) && refs.hasNonArchivedRef()
-}
-
-func (j *blockJournal) hasContext(id BlockID, context BlockContext) bool {
-	refs := j.refs[id]
-	return (refs != nil) && (refs.checkExists(context) == nil)
-}
-
-func (j *blockJournal) getDataWithContext(
-	id BlockID, context BlockContext) (
+func (j *blockJournal) getDataWithContext(id BlockID, context BlockContext) (
 	[]byte, kbfscrypto.BlockCryptKeyServerHalf, error) {
-	refEntry, err := j.getRefEntry(id, context.GetRefNonce())
-	if err != nil {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
-	}
-
-	err = refEntry.checkContext(context)
-	if err != nil {
-		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
-	}
-
-	return j.getData(id)
+	return j.s.getDataWithContext(id, context)
 }
 
-func (j *blockJournal) getAll() (
-	map[BlockID]map[BlockRefNonce]blockRefLocalStatus, error) {
-	res := make(map[BlockID]map[BlockRefNonce]blockRefLocalStatus)
-	for id, refs := range j.refs {
-		res[id] = refs.getStatuses()
-	}
-	return res, nil
+func (j *blockJournal) getUnflushedBytes() int64 {
+	// TODO: Calculate this, and possibly cache it.
```

Comment:
Okay, added some caching. The unflushed bytes now tracks all data blocks, though, even ignored ones, as keeping track of those would require more work without the initial scan. Do you think it's worth doing? This way seems more honest to the actual contents of the block journal, e.g. if there are some dangling blocks, it'd be reflected in this variable now. What do you think?